datafusion/core/src/physical_plan/file_format/parquet.rs

```diff
@@ -342,7 +379,32 @@ impl FileOpener for ParquetOpener {
                 });
 
             Ok(adapted.boxed())
-        })
+        }))
+    }
+}
+
+/// Factory of parquet file readers.
+///
+/// Provides means to implement custom data access interface.
+pub trait ParquetFileReaderFactory: Debug + Send + Sync + 'static {
+    /// Provides `AsyncFileReader` over parquet file specified in `FileMeta`
+    fn create_reader(
+        &self,
+        partition_index: usize,
+        file_meta: FileMeta,
```

Comment:
the file meta contains user defined extensions too, right? So users can pass information down into their custom readers?

Reply:
yes, exactly that. But afaik they will be able to achieve that only through custom TableProvider. Because extensions are passed through PartitionedFile into FileMeta, where higher level abstraction like `ObjectStore` produces ObjectMeta in listing operations.

Reply:
makes sense